File: LeetCode/!0238. Product of Array Except Self.cpp
/*
0238. Product of Array Except Self

思路：参考官方解法1.
先从左向右遍历，获取每个元素左边数值的乘积；
再从右向左遍历，获取每个元素右边数值的乘积；
再将两个乘积相乘，得到每个元素总乘积。
*/

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int length = nums.size();
        vector<int> L(length, 0), R(length, 0);

        L[0] = 1;
        for (int i = 1; i < length; i++) {
            L[i] = L[i - 1] * nums[i - 1];
        }

        R[length - 1] = 1;
        for (int i = length - 2; i >= 0; i--) {
            R[i] = R[i + 1] * nums[i + 1];
        }

        vector<int> result(length);
        for(int i = 0; i < length; i++) {
            result[i] = L[i] * R[i];
        }

        return result;
    }
};

File: LeetCode/!0238. Product of Array Except Self.py
'''
0238. Product of Array Except Self

给你一个长度为 n 的整数数组 nums，其中 n > 1，返回输出数组 output ，
其中 output[i] 等于 nums 中除 nums[i] 之外其余各元素的乘积。

思路：结果数组长度为n，记作res
首先，从左向右遍历数组，在res中记录每个数左边所有数字的乘积。
从右向左遍历数组，一边用right_product记录每个数字右边的所有数字乘积。
一边将其和res每个对应元素相乘。
'''

from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # 保存结果
        res = list()

        # 记录所有数左边的乘积
        res.append(1)
        for i in range(1, len(nums)):
            res.append(res[i - 1] * nums[i - 1])
          
        right_product = 1   # 记录数右边的乘积
        for i in range(len(nums) - 1, -1, -1):
            res[i] *= right_product
            right_product *= nums[i]
        
        return res
                